src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -133,64 +134,115 @@ private void CheckForShutdown()
 
-            return waitTask.IsCompletedSuccessfully
-                ? SendWithoutWaitingAsync(request, cancellationToken)
-                : WaitThenSendAsync(waitTask, request, cancellationToken);
-        }
+            TaskCompletionSourceWithCancellation<bool> waitForAvailableStreamTcs = null;
 
-        private async Task<HttpResponseMessage> WaitThenSendAsync(Task taskToWaitFor, HttpRequestMessage request, CancellationToken cancellationToken)
-        {
-            await taskToWaitFor.ConfigureAwait(false);
-            return await SendWithoutWaitingAsync(request, cancellationToken).ConfigureAwait(false);
-        }
+            lock (SyncObj)
+            {
+                long remaining = _requestStreamsRemaining;
+
+                if (remaining > 0)
+                {
+                    _requestStreamsRemaining = remaining - 1;
+                }
+                else
+                {
+                    waitForAvailableStreamTcs = new TaskCompletionSourceWithCancellation<bool>();
+                    _waitingRequests.Enqueue(waitForAvailableStreamTcs);
+                }
+            }
+
+            if (waitForAvailableStreamTcs != null)
+            {
+                await waitForAvailableStreamTcs.WaitWithCancellationAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            // Allocate an active request
 
-        private Task<HttpResponseMessage> SendWithoutWaitingAsync(HttpRequestMessage request, CancellationToken cancellationToken)
-        {
             QuicStream quicStream = null;
-            Http3RequestStream stream;
+            Http3RequestStream requestStream = null;
 
             try
             {
-                // TODO: do less work in this lock, try to get rid of goto.
                 lock (SyncObj)
                 {
-                    if (_connection == null)
+                    if (_connection != null)
                     {
-                        goto retryRequest;
+                        quicStream = _connection.OpenBidirectionalStream();
+                        requestStream = new Http3RequestStream(request, this, quicStream);
+                        _activeRequests.Add(quicStream, requestStream);
                     }
+                }
 
-                    quicStream = _connection.OpenBidirectionalStream();
-                    if (_lastProcessedStreamId != -1 && quicStream.StreamId > _lastProcessedStreamId)
-                    {
-                        goto retryRequest;
-                    }
+                if (quicStream == null)
+                {
+                    throw new HttpRequestException(SR.net_http_request_aborted, null, RequestRetryType.RetryOnSameOrNextProxy);
+                }
+
+                // 0-byte write to force QUIC to allocate a stream ID.
+                await quicStream.WriteAsync(Array.Empty<byte>(), cancellationToken).ConfigureAwait(false);
```

Comment:
😢

Reply:
Here a few thoughts I have: 
- What if we make quicStream.StreamId nullable and return null if the stream hasn't started yet?
- Why do you need to set the Http3RequestStream.StreamId here? Can you just always call into quicStream.StreamId to get the streamId?

Reply:
Finally, to track goaway, I think we may need to move this to after first write rather than here.

Reply:
> What if we make quicStream.StreamId nullable and return null if the stream hasn't started yet?

Having `StreamId` populated late is an unfortunate usability issue; I'd rather it just be usable immediately after construction if it's reasonable to do so. I think we need to better understand this from msquic perspective; my feeling is that it is behaving like this (presumably) as a corner-case optimization to avoid allocating unused stream IDs, as QUIC "allocates" gaps in stream IDs it sees and counts them against the maximum. @nibanks do you have any background here?

Reply:
Why don't you start it in your constructor then? You don't need to delay starting only when you need to send on it.

Reply:
Unless I remember incorrectly, you said that starting a stream really should be async.

Before the first write or send, there is no async operations in between. We could add a StartAsync method to the stream that's called immediately after creating the stream.